services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/device/AutoProvisioner.java

```diff
@@ -0,0 +1,416 @@
+import org.eclipse.hono.client.HonoConnection;
+import org.eclipse.hono.client.ServiceInvocationException;
+import org.eclipse.hono.client.StatusCodeMapper;
+import org.eclipse.hono.deviceregistry.service.tenant.NoopTenantInformationService;
+import org.eclipse.hono.deviceregistry.service.tenant.TenantInformationService;
+import org.eclipse.hono.deviceregistry.service.tenant.TenantKey;
+import org.eclipse.hono.service.management.Id;
+import org.eclipse.hono.service.management.OperationResult;
+import org.eclipse.hono.service.management.device.Device;
+import org.eclipse.hono.service.management.device.DeviceManagementService;
+import org.eclipse.hono.service.management.device.DeviceStatus;
+import org.eclipse.hono.tracing.TracingHelper;
+import org.eclipse.hono.util.Constants;
+import org.eclipse.hono.util.EventConstants;
+import org.eclipse.hono.util.Lifecycle;
+import org.eclipse.hono.util.MessageHelper;
+import org.eclipse.hono.util.QoS;
+import org.eclipse.hono.util.RegistrationConstants;
+import org.eclipse.hono.util.RegistryManagementConstants;
+import org.eclipse.hono.util.ResourceIdentifier;
+import org.eclipse.hono.util.TenantObject;
+import org.eclipse.hono.util.TenantResult;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import io.opentracing.Span;
+import io.opentracing.SpanContext;
+import io.opentracing.Tracer;
+import io.opentracing.noop.NoopTracerFactory;
+import io.opentracing.tag.Tags;
+import io.vertx.core.CompositeFuture;
+import io.vertx.core.Future;
+import io.vertx.core.Promise;
+import io.vertx.core.Vertx;
+import io.vertx.core.json.JsonObject;
+import io.vertx.proton.ProtonDelivery;
+
+/**
+ * Implements gateway based auto-provisioning.
+ */
+public class AutoProvisioner implements Lifecycle {
+
+    private static final Logger LOG = LoggerFactory.getLogger(AutoProvisioner.class);
+
+    private Tracer tracer = NoopTracerFactory.create();
+
+    private final DeviceManagementService deviceManagementService;
+
+    private final DeviceRegistrationInformationService deviceRegistrationInformationService;
+
+    private TenantInformationService tenantInformationService = new NoopTenantInformationService();
+
+    private DownstreamSenderFactory downstreamSenderFactory;
+
+    private Future<HonoConnection> connectionAttempt;
+
+    private Vertx vertx;
+
+    private AutoProvisionerConfigProperties config;
+
+    /**
+     * Constructs a new instance.
+     *
+     * @param deviceManagementService The device management service to be used.
+     * @param deviceRegistrationInformationService The device registration information service to be used.
+     *
+     * @throws NullPointerException if any argument is {@code null}.
+     */
+    public AutoProvisioner(
+            final DeviceManagementService deviceManagementService,
+            final DeviceRegistrationInformationService deviceRegistrationInformationService) {
```

Comment:
I don't quite get why some (required) collaborators are passed in via the constructor while others are injected by means of a setter. Is there a pattern that I don't see?
Why do we need a dedicated `DeviceRegistrationInformationService`? Can't we simply use the existing `RegistrationService` for this purpose?

Reply:
No, actually there is no pattern. Refactored it to using setters.

I could not find a method in `RegistrationService` which just gets the registration data of a device. That's also implied by their names since they start with "assert..." - these methods also query the tenant service which is unnecessary overhead when the intention is just to get the registration data.

Also I thought `RegistrationService` should represent the Device Registration API. Since there is no simple "get" method defined there and its only for internal use I did not want to extend the API. Also I did want to add the method there in order no to mix up "API definitions" and "internal helpers".

And, finally 😅, I got the impression that `TenantInformationService` was defined because of similar reasons.

Reply:
well, the `TenantInformationService` has been created because there are deployment environments where Tenants will not be managed as part of the device registry and will instead be represented by e.g. custom resources in a kubernetes cluster.
I might not fully understand what the `DeviceRegistrationInformationService.processAssertRegistration` method is actually returning. The JavaDoc ist quite vaguely referring to *the information registered for the device*.

I wonder why the method is named *processAssertRegistration* if it is not about performing the assertion of the device's registration status. if it is about retrieving the device's registration information then why isn't it called *getRegistrationInformation*?

it would be helpful if the JavaDoc would define more precisely what data is actually being returned by the method. You might be able to refer to the *Device* object definition in the Management API?

Reply:
It is still named _processAssertRegistration_ because I tried to keep the PR as small as possible, see https://github.com/eclipse/hono/pull/2094#discussion_r540762386 

I am happy to change it though, since two reviewers have now stumbled upon it - I guess the change is still "digestible" then.

> it would be helpful if the JavaDoc would define more precisely what data is actually being returned by the method. You might be able to refer to the Device object definition in the Management API?

Yes, that makes sense.